### PR TITLE
Refactor PR description building and write tests

### PR DESCRIPTION
**Pull request by dkurepa**
Issue link : https://github.com/dotnet/arcade/issues/7339

### CODE REVIEW

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (premun) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:884]: nit: You could also turn `startingReferenceId` into a `ref` argument and make sure it gets incremented properly as part of `CalculatePRDescription` so that it's more clear you need to pass this along between the calls

**Review by premun — APPROVED** [review]
Nice! Looks good, we should monitor it closely after merging to see if we managed to fix it

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (lpatalas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:14]: I think this class would have a bit nicer interface if the `StringBuilder` was created internally by this class. I see two issues with the current design:
- This class does not control the `StringBuilder` so other code can do something unexpected with it between the calls to `CalculatePRDescription` like clearing it for example
- For someone reading the code it may look a bit magical because you have to be aware that the output is a side effect of writing to `description` that you passed as constructor argument. If the result was built internally you would have to get it in more explicit way, e.g. (pseudocode):

```csharp
var builder = new PullRequestDescriptionBuilder(loggerFactory);
foreach (var commit in commits)
{
    builder.CalculatePRDescription(...);
}

// New method that just returns stringBuilder.ToString() for private builder
string result = builder.GetDescription();

// then you can append it to other builder
description.Append(result);
```

I think this would make it easier to grasp what is happening when you're reading the code.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (lpatalas) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:872]: Nitpick: variable should be lowercase
```suggestion
            PullRequestDescriptionBuilder pullRequestDescriptionBuilder = new PullRequestDescriptionBuilder(LoggerFactory, description);
```

**Review by lpatalas — COMMENTED** [review]
LGTM apart from a few minor comments.

File: src/Maestro/tests/SubscriptionActorService.Tests/PullRequestDescriptionBuilderTests.cs
Comment (premun) [reply src/Maestro/tests/SubscriptionActorService.Tests/PullRequestDescriptionBuilderTests.cs:25]: FYI - there is also `Mock.Of<T>()` which gives you the mock directly so no need for `new` and `.Object`.

In this call, you can also setup properties right away:
```csharp
T x = Mock.Of<T>(m => m.Prop1 == "value1" && m.Prop2 == 33);
```

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (premun) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:872]: Also we should do `var` or 
```
PullRequestDescriptionBuilder pullRequestDescriptionBuilder = new(LoggerFactory, description);
```
?

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:14]: I think this also lets us make this a static helper class. This class doesn't seem to store any data that it needs per instance. This class' public interface right now is just `CalculatePRDescriptionAsync`, and that method gets all the data it needs not through instance variables, but from external data from the caller so it can be a static method that also takes the existing PR description and the logger as parameters.

Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:15]: I don't think this is true anymore, I didn't consider the case where we update an existing PR for which we need to calculate the startingReferenceId based on the previous ones, so the class has data

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs]: I moved it inside the class

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs]: I checked and it's all mixed, we use var in some places and not in others

Comment (premun) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs]: I think we use var in "obvious" places like this one where we repeat the type for no reason `T x = new T();`
and keep it in places where it's not obvious `T x = some.method();`

However, we seem to not prefer var based on `.editorconfig` - https://github.com/dotnet/arcade/blob/fc232204747e6249e73b11e27aeec4de58ab00c1/.editorconfig#L53

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (premun) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs]: Even better

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:160]: [nit] include a comment on what you're looking to capture here.

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:166]: [nit] I recommend using `Math.Max`, or you can transform this whole foreach into a LINQ expression using `Max`.

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:31]: [nit] docs need to be updated with the new parameter.

File: src/Maestro/tests/SubscriptionActorService.Tests/PullRequestDescriptionBuilderTests.cs
Comment (riarenas) [inline src/Maestro/tests/SubscriptionActorService.Tests/PullRequestDescriptionBuilderTests.cs:18]: [nit] Seems like these could all be private

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:15]: yep, with the startingReferenceId it makes sense to not make it just static. But I think @lpatalas comments about how this class has undocumented side effects in constructor arguments still apply. If I instantiate a PullRequestDescriptionBuilder, I would expect to have to interact with it to get a PullRequestDescription, but instead everything is done to a string builder that I passed to the constructor.

Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs]: I refactored the code a bit by removing the StringBuilders from the outside of PullRequestDescriptionBuilder. The descriptions are now moved around with return values. This should remove the ability for something unexpected to happen to the StringBuilders

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:196]: Is there any reason not to make CalculatePRDescription return this instead of being void? Nothing stops a client from calling this method without calling Calculate first, only to find out they got an unexpected result.

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:165]: [Nit]:
```suggestion
            //example: given [23]:sometext as input, it will attempt to capture "23"
```

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:16]: [nit] we should add documentation for this constructor, especially to make it clear what the description string is supposed to be.

**Review by riarenas — APPROVED** [review]
Just a couple of nits

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:196]: we'd be returning incomplete PR descriptions and overwriting them for every update other than the last one, which would be complete

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:196]: I don't think I follow. A call to calculatePRDescription gives you a full description for the PR given a new set of updates and a new build, doesn't it? My expectation when using this class would be something like this:

```c#
string newDescription = prDescriptionBuilder.CalculatePRDescription(params...)
```

By adding this method, you run the real risk of making mistakes like:

```c#
var prDescriptionBuilder = new PullRequestDescriptionBuilder(loggerFactory, string.empty)
string description = prDescriptionBuilder.GetPRDescription() // empty string? I thought I was getting a description!
```

and nothing is actually letting you know that you used the class wrong.

Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:199]: The way it works now we call prDescriptionBuilder.CalculatePRDescription() for every update. So we only get the complete descrption after the last update. The CommitUpdatesAsync in the PullRequestActor.cs does this

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:199]: That means this class needs a very particular way of using it, which is not made clear by the public interface of the class. We can't assume that every client of this class will have the same calling pattern. 

* Nothing stops me right now from providing an empty description when initializing
* Nothing stops me from calling GetPRDescription before the full description has been calculated

By merging the two methods, you ensure that you always get a partial description for that point in time.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: I missed this change, and I don't agree with it. I don't think that method should do more than what it says on the name. Which is to commit the changes.

**Review by riarenas — CHANGES_REQUESTED** [review]
I don't think CommitUpdatesAsync should be changed to return a PR description.

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:17]: What does initialized PR description mean?

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: But this method was calculating the PR description before, it just passed it to the caller via the StringBuilder

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:17]: In case of a new PR the CreatePullRequestAsync Adds some lines to the description ("initializes it") and then passes it to CommitUpdateAsync which calls the PRBuilder

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: Maybe we can change the naming a bit?

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (dkurepa) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:199]: I think maybe that the CalculatePRDescription isn't a very good name for the method, as it implies that it is called only once. But this is not the case we have to call this method for every repo separately. Maybe something like AppendPRDescription() would be a better name for this method, and we could rename the GetPRDescrption to ToString() or something like that

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: Yep I think if it comes with a name change this makes some more sense

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:17]: Do you think it makes sense to make the description an optional parameter to the constructor?

* If the description is null or empty, initialize it with the introductory text.
* If it already has something, the class assumes it's an already properly formatted string. 

It seems a little weird that this class' responsibility is to create pull request descriptions, but it needs a string that already represents a pr description.

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:199]: I like that. Maybe call it AppendBuild? Since it's appending the updates from a new build?

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:17]: This would allow us to move https://github.com/dotnet/arcade-services/pull/1774/files#diff-c5bf57a757adcea620af121a78f82cd95fec509a1de1ebc7f63a4a7900d3d02aR757-R759 to the constructor for this class.

File: src/Maestro/SubscriptionActorService/PullRequestActor.cs
Comment (lpatalas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: I agree with Ricardo but it was also like this before. We were just sneaking it outside in a `StringBuilder` :) The simplest refactoring here would be to introduce some helper class `UpdateCommitResult` with one property called `PRDescription` and return this instead of bare string. It won't be a revolution but it would at least make explicit what we are returning.

Comment (riarenas) [reply src/Maestro/SubscriptionActorService/PullRequestActor.cs:851]: I'm totally fine with renaming this to CalculatePRDescriptionAndCommitUpdatesAsync to at least make it explicit. Of course bonus points if we can keep improving things as part of the refactoring, but I don't want to block on this since we were already doing it.

**Review by riarenas — APPROVED** [review]
I'm basically back to just nits after the discussion, so I don't want to block the PR on them. LGTM if we apply some of the changes we discussed. Thanks @dkurepa!

File: src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs
Comment (premun) [inline src/Maestro/SubscriptionActorService/PullRequestDescriptionBuilder.cs:29]: Make all fields `readonly` (where possible)

```suggestion
        private readonly ILogger _logger;

        private readonly StringBuilder _description;
```

**Review by lpatalas — APPROVED** [review]
Looks good after the changes. Approved :)

**Pull request merged**
